AWS/connectors/__init__.py

```diff
@@ -0,0 +1,98 @@
+"""All available connectors for this module."""
+import asyncio
+import time
+from abc import ABCMeta
+from functools import cached_property
+
+from pydantic import BaseModel, Field
+from sekoia_automation.aio.connector import AsyncConnector
+from sekoia_automation.aio.helpers.aws.client import AwsClient, AwsConfiguration
+from sekoia_automation.connector import DefaultConnectorConfiguration
+from sekoia_automation.module import Module
+
+from .metrics import EVENTS_LAG, FORWARD_EVENTS_DURATION, OUTCOMING_EVENTS
+
+
+class AwsModuleConfiguration(BaseModel):
+    """The configuration of the AWS module."""
+
+    aws_access_key: str = Field(..., description="The identifier of the access key")
+    aws_secret_access_key: str = Field(secret=True, description="The secret associated to the access key")
+    aws_region_name: str = Field(..., description="The area hosting the AWS resources")
+
+
+class AwsModule(Module):
+    """The AWS module."""
+
+    configuration: AwsModuleConfiguration
+
+
+class AbstractAwsConnector(AsyncConnector, metaclass=ABCMeta):
+    """The abstract connector."""
+
+    module: AwsModule
+    configuration: DefaultConnectorConfiguration
+
+    @cached_property
+    def aws_client(self) -> AwsClient:
+        """
+        Base implementation of AWS client.
+
+        AwsClient contains `get_client` method with correct initialization.
+
+        Returns:
+            AwsClientT:
+        """
+        config = AwsConfiguration(
+            aws_access_key_id=self.module.configuration.aws_access_key,
+            aws_secret_access_key=self.module.configuration.aws_secret_access_key,
+            aws_region=self.module.configuration.aws_region_name,
+        )
+
+        return AwsClient(config)
+
+    async def next_batch(self) -> list[str]:
+        """
+        Get next batch of messages.
+
+        Contains main logic of the connector.
+
+        Returns:
+            list[str]:
+        """
+        raise NotImplementedError("next_batch method must be implemented")
+
+    def run(self) -> None:  # pragma: no cover
+        """Run the connector."""
+        loop = asyncio.get_event_loop()
+
+        previous_processing_end = None
+        try:
+            while True:
+                processing_start = time.time()
+                if previous_processing_end is not None:
+                    EVENTS_LAG.labels(intake_key=self.configuration.intake_key).observe(
+                        processing_start - previous_processing_end
+                    )
```

Comment:
The SQS [ReceiveMessage](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ReceiveMessage.html) action allow you to get the `SentTimestamp` attribute name. I wonder if using this attribute to calculate the delay is more appropriate. Could you take a look?
Thanks

Reply:
If I'm correct the main goal of `EVENTS_LAG` is to check delay between runs in connector.
As per official docs `SentTimestamp` is the time when the message was sent to the queue.
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ReceiveMessage.html
Lets discuss in DM